docs/nddata/subclassing.rst

```diff
@@ -1,100 +1,487 @@
+.. _nddata_subclassing:
+
 Subclassing
 ===========
 
-There are a couple of choices to be made in subclassing from the nddata
-package. For the greatest flexibility, subclass from
-`~astropy.nddata.NDDataBase`, which places no restrictions on any of
-its attributes. In many cases, subclassing `~astropy.nddata.NDData` will work
-instead; it is more straightforward but places some minimal restrictions on
-how the data can be represented.
+`~astropy.nddata.NDData`
+------------------------
+
+This class serves as the base for subclasses that use a `numpy.ndarray` (or
+something that presents a numpy-like interface) as the ``data`` attribute.
+
+.. note::
+  Each attribute is saved as attribute with one leading underscore. For example
+  the ``data`` is saved as ``_data`` and the ``mask`` as ``_mask``, and so on.
+
+Adding another property
+^^^^^^^^^^^^^^^^^^^^^^^
+
+    >>> from astropy.nddata import NDData
+
+    >>> class NDDataWithFlags(NDData):
+    ...     def __init__(self, *args, **kwargs):
+    ...         # Remove flags attribute if given and pass it to the setter.
+    ...         self.flags = kwargs.pop('flags') if 'flags' in kwargs else None
+    ...         super(NDDataWithFlags, self).__init__(*args, **kwargs)
+    ...
+    ...     @property
+    ...     def flags(self):
+    ...         return self._flags
+    ...
+    ...     @flags.setter
+    ...     def flags(self, value):
+    ...         self._flags = value
+
+    >>> ndd = NDDataWithFlags([1,2,3])
+    >>> ndd.flags is None
+    True
+
+    >>> ndd = NDDataWithFlags([1,2,3], flags=[0, 0.2, 0.3])
+    >>> ndd.flags
+    [0, 0.2, 0.3]
+
+.. note::
+  To simplify subclassing each setter (except for ``data``) is called during
+  ``__init__`` so putting restrictions on any attribute can be done inside
+  the setter and will also apply duing instance creation.
+
+Customize the setter for a property
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    >>> import numpy as np
+
+    >>> class NDDataMaskBoolNumpy(NDData):
+    ...
+    ...     @NDData.mask.setter
+    ...     def mask(self, value):
+    ...         # Convert mask to boolean numpy array.
+    ...         self._mask = np.array(value, dtype=np.bool_)
+
+    >>> ndd = NDDataMaskBoolNumpy([1,2,3])
+    >>> ndd.mask = True
```

Comment:
Since #4879 is merged the ``mask`` could also be given during ``__init__``. Should I rebase and add this as example? Or should I mention which setters are called during ``__init__`` (wouldn't need a rebase).